// @aries-framework/anoncreds/build/repository/AnonCredsCredentialDefinitionRecord.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AnonCredsCredentialDefinitionRecord = void 0;
const core_1 = require("@aries-framework/core");
class AnonCredsCredentialDefinitionRecord extends core_1.BaseRecord {
    constructor(props) {
        var _a;
        super();
        this.type = AnonCredsCredentialDefinitionRecord.type;
        if (props) {
            this.id = (_a = props.id) !== null && _a !== void 0 ? _a : core_1.utils.uuid();
            this.credentialDefinitionId = props.credentialDefinitionId;
            this.credentialDefinition = props.credentialDefinition;
        }
    }
    getTags() {
        return Object.assign(Object.assign({}, this._tags), { credentialDefinitionId: this.credentialDefinitionId, schemaId: this.credentialDefinition.schemaId, issuerId: this.credentialDefinition.issuerId, tag: this.credentialDefinition.tag });
    }
}
exports.AnonCredsCredentialDefinitionRecord = AnonCredsCredentialDefinitionRecord;
AnonCredsCredentialDefinitionRecord.type = 'AnonCredsCredentialDefinitionRecord';
