// @aries-framework/indy-sdk/build/anoncreds/utils/transform.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.indySdkRevocationDeltaFromAnonCreds = exports.indySdkRevocationRegistryFromAnonCreds = exports.anonCredsRevocationStatusListFromIndySdk = exports.indySdkRevocationRegistryDefinitionFromAnonCreds = exports.indySdkCredentialDefinitionFromAnonCreds = exports.anonCredsCredentialDefinitionFromIndySdk = exports.indySdkSchemaFromAnonCreds = exports.anonCredsSchemaFromIndySdk = void 0;
const identifiers_1 = require("./identifiers");
function anonCredsSchemaFromIndySdk(schema) {
    const { did } = (0, identifiers_1.parseSchemaId)(schema.id);
    return {
        issuerId: did,
        name: schema.name,
        version: schema.version,
        attrNames: schema.attrNames,
    };
}
exports.anonCredsSchemaFromIndySdk = anonCredsSchemaFromIndySdk;
function indySdkSchemaFromAnonCreds(schemaId, schema, indyLedgerSeqNo) {
    return {
        id: schemaId,
        attrNames: schema.attrNames,
        name: schema.name,
        version: schema.version,
        ver: '1.0',
        seqNo: indyLedgerSeqNo,
    };
}
exports.indySdkSchemaFromAnonCreds = indySdkSchemaFromAnonCreds;
function anonCredsCredentialDefinitionFromIndySdk(credentialDefinition) {
    const { did } = (0, identifiers_1.parseCredentialDefinitionId)(credentialDefinition.id);
    return {
        issuerId: did,
        schemaId: credentialDefinition.schemaId,
        tag: credentialDefinition.tag,
        type: 'CL',
        value: credentialDefinition.value,
    };
}
exports.anonCredsCredentialDefinitionFromIndySdk = anonCredsCredentialDefinitionFromIndySdk;
function indySdkCredentialDefinitionFromAnonCreds(credentialDefinitionId, credentialDefinition) {
    return {
        id: credentialDefinitionId,
        schemaId: credentialDefinition.schemaId,
        tag: credentialDefinition.tag,
        type: credentialDefinition.type,
        value: credentialDefinition.value,
        ver: '1.0',
    };
}
exports.indySdkCredentialDefinitionFromAnonCreds = indySdkCredentialDefinitionFromAnonCreds;
function indySdkRevocationRegistryDefinitionFromAnonCreds(revocationRegistryDefinitionId, revocationRegistryDefinition) {
    return {
        id: revocationRegistryDefinitionId,
        credDefId: revocationRegistryDefinition.credDefId,
        revocDefType: revocationRegistryDefinition.revocDefType,
        tag: revocationRegistryDefinition.tag,
        value: {
            issuanceType: 'ISSUANCE_BY_DEFAULT',
            maxCredNum: revocationRegistryDefinition.value.maxCredNum,
            publicKeys: revocationRegistryDefinition.value.publicKeys,
            tailsHash: revocationRegistryDefinition.value.tailsHash,
            tailsLocation: revocationRegistryDefinition.value.tailsLocation,
        },
        ver: '1.0',
    };
}
exports.indySdkRevocationRegistryDefinitionFromAnonCreds = indySdkRevocationRegistryDefinitionFromAnonCreds;
function anonCredsRevocationStatusListFromIndySdk(revocationRegistryDefinitionId, revocationRegistryDefinition, delta, timestamp, isIssuanceByDefault) {
    var _a, _b;
    // 0 means unrevoked, 1 means revoked
    const defaultState = isIssuanceByDefault ? 0 : 1;
    // Fill with default value
    const revocationList = new Array(revocationRegistryDefinition.value.maxCredNum).fill(defaultState);
    // Set all `issuer` indexes to 0 (not revoked)
    for (const issued of (_a = delta.value.issued) !== null && _a !== void 0 ? _a : []) {
        revocationList[issued] = 0;
    }
    // Set all `revoked` indexes to 1 (revoked)
    for (const revoked of (_b = delta.value.revoked) !== null && _b !== void 0 ? _b : []) {
        revocationList[revoked] = 1;
    }
    return {
        issuerId: revocationRegistryDefinition.issuerId,
        currentAccumulator: delta.value.accum,
        revRegId: revocationRegistryDefinitionId,
        revocationList,
        timestamp,
    };
}
exports.anonCredsRevocationStatusListFromIndySdk = anonCredsRevocationStatusListFromIndySdk;
function indySdkRevocationRegistryFromAnonCreds(revocationStatusList) {
    return {
        ver: '1.0',
        value: {
            accum: revocationStatusList.currentAccumulator,
        },
    };
}
exports.indySdkRevocationRegistryFromAnonCreds = indySdkRevocationRegistryFromAnonCreds;
function indySdkRevocationDeltaFromAnonCreds(revocationStatusList) {
    // Get all indices from the revocationStatusList that are revoked (so have value '1')
    const revokedIndices = revocationStatusList.revocationList.reduce((revoked, current, index) => (current === 1 ? [...revoked, index] : revoked), []);
    return {
        value: {
            accum: revocationStatusList.currentAccumulator,
            issued: [],
            revoked: revokedIndices,
            // NOTE: I don't think this is used?
            prevAccum: '',
        },
        ver: '1.0',
    };
}
exports.indySdkRevocationDeltaFromAnonCreds = indySdkRevocationDeltaFromAnonCreds;
